controller: Add overhead annotation using pod mutating webhook

c3d opened the pull request:
In order for the Kata runtime to correctly deal with hot plugging and set a VM size that is compatible with the host cgroup size [1], we need to pass the known overhead (from the RuntimeClass) to the runtime using an annotation.

This code implements a mutating web hook to modify the pod so that the annotation is added when the runtime class is Kata.

Fixes: #1057 

**- Description of the problem which is fixed/What is the use case**

When hot-plugging memory in Kata (typically because the container spec provides a memory limit), the host-side cgroup and the maximum VM size currently become inconsistent, because the host-side cgroup is defined as "pod overhead + container limit" whereas the actual maximum VM size is "default VM size + container limit".

The in-guest cgroup is insufficient to constrain the memory usage sufficiently, because it does not account for things like kernel buffers. This leads to situations where we end up hitting a host-side cgroup out-of-memory, which kills the VM in a nasty way: we lose connexion to the kata runtime itself, and therefore things like container exit status, etc.

**- What I did**

This adds a pod mutating web hook that injects an annotation that [the runtime can use](https://github.com/kata-containers/kata-containers/pull/11824) to adjust the amount of memory being hot-plugged to avoid the issue.

**- How to verify it**

There is some local testing of the operator's pod mutation logic which can be exercised by running the run_pod_overhead_mutators.sh

**- Description for the changelog**

Add pod-mutating web hook to inject memory overhead annotation for the kata runtime

Assisted-by: Cursor with gpt5 model (test generation)

c3d committed:
controller: Add overhead annotation using pod mutating webhook

In order for the Kata runtime to correctly deal with hot plugging and
set a VM size that is compatible with the host cgroup size [1], we
need to pass the known overhead (from the RuntimeClass) to the runtime
using an annotation.

This code implements a mutating web hook to modify the pod so that the
annotation is added when the runtime class is Kata.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d commented:
I don't have the proper privileges to change labels on this repo, but this is still work-in-progress. The actual interaction with the runtime remains to be tested.

gkurz commented on api/v1/kataconfig_webhook.go:
```diff
@@ -61,16 +61,28 @@ func (r *KataConfig) ValidateCreate(ctx context.Context, obj runtime.Object) (ad
+		}
+		if *r.Spec.MemoryOverheadMB > 4096*1024 {
+			return nil, fmt.Errorf("memoryOverheadMB must be at most 4GB")
+		}
```
Can you clarify how you came up with these numbers ?

c3d replied:
Added comments:
```
                // The Linux kernel was tested not to boot reliably with less than 60MB                                                                                                                                                                                  
                if *r.Spec.MemoryOverheadMB < 60 {                              │
                        return nil, fmt.Errorf("memoryOverheadMB must be at least 60MB")
                }                                                               │
                // Kata configures the VM with 2G by default, so overhead twice as big is likely an error                                                                                                                                                                
                if *r.Spec.MemoryOverheadMB > 4096*1024 {                       │
                        return nil, fmt.Errorf("memoryOverheadMB must be at most 4GB")
                }                                                               │
```

gkurz commented on api/v1/kataconfig_webhook.go:
```diff
@@ -85,6 +97,15 @@ func (r *KataConfig) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.
+			return nil, fmt.Errorf("memoryOverheadMB must be at most 4GB")
+		}
+	}
+
```
Can you move the duplicate code block to some helper ?

c3d replied:
Done

gkurz commented on config/crd/bases/kataconfiguration.openshift.io_kataconfigs.yaml:
```diff
@@ -124,6 +124,12 @@ spec:
+                default: 350
+                format: int32
+                maximum: 2048
+                minimum: 1
```
Why do the min and max don't match what's tested in the code above ?

c3d replied:
Oversight.

gkurz commented on config/manager/manager.yaml:
```diff
@@ -86,7 +97,7 @@ spec:
               value: registry.redhat.io/openshift-sandboxed-containers/osc-podvm-builder-rhel9@sha256:517b1d76dad553871c745a315aa73c341f4672029704d50992a3eb663945e6dc
             - name: RELATED_IMAGE_PODVM_PAYLOAD
-              value: registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9@sha256:211a4291cd13302a18ae2766a8e680a2593450ce31f119e78ac3367c5858104a 
+              value: registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9@sha256:211a4291cd13302a18ae2766a8e680a2593450ce31f119e78ac3367c5858104a
```
Current value in the `devel` branch is :

```
registry.redhat.io/openshift-sandboxed-containers/osc-podvm-payload-rhel9@sha256:9c07280746952c7bd4ea811cd15e6dc2e35e0f6c2015e21436dfeaa418ec5799 
```

More generally,  these lines with image digests are _maintained_ by mintmaker. You should never have to modify them in a PR, so please just drop this change.

This being said, there's indeed a trailing space... even in the current HEAD of `devel`, Cc @littlejawa .

gkurz replied:
It seems that this space has been here for a very long time and mintmaker PRs gladly honor this formatting nit :wink:

gkurz replied:
Let's fix that in a separate PR.

gkurz commented on controllers/pod_mutator.go:
```diff
@@ -0,0 +1,186 @@
+
+	// Known Kata runtime classes
+	KataRuntimeClass       = "kata"
+	KataRemoteRuntimeClass = "kata-remote"
```
No need to duplicate strings that we already have in the main controller file. The controller maintains a list of the known kata runtime class in the KataConfig.

https://github.com/openshift/sandboxed-containers-operator/blob/devel/api/v1/kataconfig_types.go#L51

gkurz commented on controllers/pod_mutator_test.go:
```diff
@@ -0,0 +1,802 @@
+					Namespace: "default",
+				},
+				Spec: corev1.PodSpec{
+					RuntimeClassName: stringPtr("kata-remote"),
```
Maybe use https://github.com/openshift/sandboxed-containers-operator/blob/devel/controllers/peerpods.go#L45 ?

And fix other sites.

gkurz replied:
You might need to rebase to get the changes from #1069.

gkurz commented on controllers/pod_mutator_test.go:
```diff
@@ -0,0 +1,802 @@
+					Namespace: "default",
+				},
+				Spec: corev1.PodSpec{
+					RuntimeClassName: stringPtr("kata"),
```
Maybe use https://github.com/openshift/sandboxed-containers-operator/blob/devel/controllers/openshift_controller.go#L78 ?

And fix other sites.

gkurz commented on api/v1/kataconfig_types.go:
```diff
@@ -44,6 +44,12 @@ type KataConfigSpec struct {
 	EnablePeerPods bool `json:"enablePeerPods"`
+
+	// +optional
+	// +kubebuilder:default:=350
```
Is this taken from https://github.com/openshift/sandboxed-containers-operator/blob/devel/controllers/openshift_controller.go#L81 ?

gkurz replied:
If the answer is yes, should they stay in sync ?

gkurz commented:
> I don't have the proper privileges to change labels on this repo, but this is still work-in-progress. The actual interaction with the runtime remains to be tested.

Converting the PR to draft is the way to go.

c3d committed:
tests: Move test scripts under scripts/webhook-tests

Add two test scripts
- run_pod_overhead_mutator.sh that does a local unit test run of the
  webhook and pod overhead mutator
- test_memory_overhead is an end-to-end test for the pod mutator

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>
Assisted-by: claude-4.5-sonnet

c3d committed:
docs: Add documentation for pod-mutating webhook

Add a proposal explaining the rationale and implementation, and a
testing guide describing how this can be tested.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>
Assisted-by: claude-4.5-opus-high

gkurz commented on config/certmanager/certificate.yaml:
```diff
@@ -6,15 +6,15 @@ apiVersion: cert-manager.io/v1
 metadata:
   name: selfsigned-issuer
-  namespace: system
+  namespace: openshift-sandboxed-containers-operator
```
What's the reason for that ? This comes from Operator SDK AFAIK. Not sure it should be changed...

c3d replied:
This was for testing purpose, should not be committed. Dropping it.

gkurz commented on config/manager/manager.yaml:
```diff
@@ -12,7 +12,7 @@ apiVersion: apps/v1
 metadata:
   name: controller-manager
-  namespace: system
+  namespace: openshift-sandboxed-containers-operator
```
Same remark. This is rendered with `openshift-sandboxed-containers-operator` when updating the bundle.

c3d replied:
Same reason, reverting that change too.

gkurz commented on config/manager/manager.yaml:
```diff
@@ -55,30 +55,41 @@ spec:
                 name: peer-pods-cm
                 optional: true
-          image: controller:latest
+          image: quay.io/c3d/openshift-sandboxed-containers-operator:v1.10.1-memory-overhead
```
Same here, this is rendered in the final bundle. And we're not going to merge a reference to your personal repo, do we ? :wink:

c3d replied:
No, all this was for testing purpose and should not have been committed ;-)

bpradipt commented on api/v1/kataconfig_types.go:
```diff
@@ -44,6 +44,12 @@ type KataConfigSpec struct {
+	// +kubebuilder:default:=350
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=2048
+	MemoryOverheadMB *int32 `json:"memoryOverheadMB,omitempty"`
```
Is updating CRD really needed? 
We have consciously avoided updating KataConfig CRD till now and instead using feature configmap. 
If we are updaing the CRD then we need to handle migration for installed CRDs and probably version it as well.

c3d replied:
This is a very valid point. I will rework

c3d committed:
Move pod overhead control to a feature gate

In order to avoid changing the KataConfig CRD, move the corresponding
settings to a feature gate

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
webhook: Remove duplicate pod mutator registration

Without this fix, I would get:

```
export KUBECONFIG=/home/ddd/k8s/kubeconfig-v718-ocp420
oc -n openshift-sandboxed-containers-operator logs deploy/controller-manager
2026-02-12T10:40:00Z	INFO	controller-runtime.webhook	Registering webhook	{"path": "/mutate-pods-v1"}
panic: can't register duplicate path: /mutate-pods-v1
```

So need to remove duplicate registration

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
test_memory_overhead: Remove invalid option for `oc debug`

The `oc debug` step was failing:

```
   error: unknown shorthand flag: 'i' in -it
   See 'oc debug --help' for usage.
```

Remove the flag. After that the tests pass.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

bpradipt commented:
With K8s 1.34 there is MutatingAdmissionPolicy which provides an in-process alternative compared to a mutating webhook.
https://kubernetes.io/docs/reference/access-authn-authz/mutating-admission-policy/

Just to keep in mind going forward

c3d committed:
Avoid a crash in the operator if discovery fails

Under some conditions, connectivity may fail in the IsOpenShift()
function, which causes the operator to crash repeatedly.
(This may actually result from the cluster being unhealthy)

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
doc: Add documentation and samples for how to use the feature

Add a new directory, config/samples/pod-overhead-annotation, detailing
how to use the new feature, along with a few sample config files.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
agents: Add instructions for agents based on developer's guide

Add some instructions to the agents based on the recently proposed
developer's guide (#1760).

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
test: add script to align webhook TLS secret with cert-manager CA

Find Deployment volume webhook-cert by name and set secretName to
webhook-server-cert so apiserver caBundle matches the served certificate on
OLM-shaped clusters.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>
Made-with: Cursor

c3d committed:
controller: Fix crash in IsOpenShift

`IsOpenShift()` now probes `ServerResourcesForGroupVersion` for
`"config.openshift.io/v1"`` instead of `ServerGroups()`, because
aggregated discovery can still fail when `metrics.k8s.io` is
unhealthy, which manifests as a CrashLoop until a new image is built.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
webhook: Configuration controller-manager-service

Needed certificates for the web hook

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

c3d committed:
tests: Add support for runtime-rs test

Add support to the pod mutator in the controller and to the E2E test
script to test the runtime-rs for annotation limitation.

Signed-off-by: Christophe de Dinechin <dinechin@redhat.com>

gkurz commented:
@c3d FYI there's some ongoing work in this area upstream but it is for runtime-rs only.

https://github.com/kata-containers/kata-containers/pull/12735
